core-bundle/src/Controller/SitemapController.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Controller;
+
+use Contao\Backend;
+use Contao\CoreBundle\Event\ContaoCoreEvents;
+use Contao\CoreBundle\Event\SitemapEvent;
+use Contao\PageModel;
+use Contao\System;
+use Contao\User;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
+
+/**
+ * @Route(defaults={"_scope" = "frontend"})
+ *
+ * @internal
+ */
+class SitemapController extends AbstractController
+{
+    /**
+     * @Route("/sitemap.xml")
+     */
+    public function __invoke(Request $request): Response
+    {
+        $this->initializeContaoFramework();
+
+        /** @var PageModel $pageModel */
+        $pageModel = $this->get('contao.framework')->getAdapter(PageModel::class);
+
+        $rootPages = $pageModel->findPublishedRootPages(['dns' => $request->server->get('HTTP_HOST')]);
+
+        if (null === $rootPages) {
+            return new Response('', Response::HTTP_NOT_FOUND);
+        }
+
+        $pages = [];
+        $rootPageIds = [];
+        $tags = ['contao.sitemap'];
+
+        foreach ($rootPages as $rootPage) {
+            /** @var Backend $backend */
+            $backend = $this->get('contao.framework')->getAdapter(Backend::class);
+            $pages = array_merge($pages, $backend->findSearchablePages($rootPage->id, '', true));
```

Comment:
If we can, we should avoid using `array_merge()` inside a foreach loop (see https://3v4l.org/ThGtv).

Reply:
I don't think we can.

Reply:
Yes, we can. Just click the link above. 😉

Reply:
I don't think this will work with https://github.com/contao/contao/pull/2295#discussion_r503124297.

Reply:
Ok.